File: ryan/yilma/deleteFace.php
<?php
session_start();
header('Content-Type: text/plain');

require_once '../database.php';
$db = new database();
$db->ensureAdminTables();

if (!isset($_SESSION['valid_user']) || !$db->isAdmin($_SESSION['valid_user'])) {
    http_response_code(403);
    echo 'Unauthorized';
    exit;
}

$perm = $db->getUserPermissions($_SESSION['valid_user']);
$canManageFaces = ((int)$perm['can_manage_faces'] === 1) || $db->isAdmin($_SESSION['valid_user']);
if (!$canManageFaces) {
    http_response_code(403);
    echo 'Permission denied';
    exit;
}

if (!isset($_POST['user_id'])) {
    http_response_code(400);
    echo 'Missing user_id';
    exit;
}

$user_id_raw = trim($_POST['user_id']);
if ($user_id_raw === '') {
    http_response_code(400);
    echo 'Empty user_id';
    exit;
}

$python = '/var/www//py311env/bin/python3.11';
$script = '/var/www/html/htdocs/ryan/yilma/deleteFace.py';
$cmd = 'sudo ' . $python . ' ' . escapeshellarg($script) . ' ' . escapeshellarg($user_id_raw);

$output = [];
$return_code = 0;
exec($cmd . ' 2>&1', $output, $return_code);
$body = implode("\n", $output);

$actor = $_SESSION['valid_user'];
$ip = isset($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : null;

if ($return_code === 0) {
    $db->logAdminEvent($actor, 'face_deleted', $user_id_raw, 'Deleted face data and retraining started', $ip);
    echo "Deleted and retraining started.\n" . $body . "\n";
    exit;
}

if ($return_code === 1) {
    $db->logAdminEvent($actor, 'face_delete_failed', $user_id_raw, 'No exact face directory match', $ip);
    http_response_code(404);
    echo "No exact match for user directory. Case-sensitive.\n" . $body . "\n";
    exit;
}

$db->logAdminEvent($actor, 'face_delete_failed', $user_id_raw, 'Deletion script error', $ip);
http_response_code(500);
echo "Error during deletion.\n" . $body . "\n";

File: ryan/yilma/deleteFace.py
#!/usr/bin/env python3.11
# -*- coding: utf-8 -*-
import os
import shutil
import subprocess
import sys
import traceback
import unicodedata

BASE_DIR = "/var/www/html/htdocs/ryan/yilma"
TRAINING_ROOT = f"{BASE_DIR}/training_images"
TRAINER_SCRIPT = f"{BASE_DIR}/trainer.py"
MAX_SAMPLE_DIRS = 20


def normalize_user_id(user_id: str) -> str:
    return unicodedata.normalize("NFKC", (user_id or "").strip())


def list_training_dirs():
    try:
        entries = os.listdir(TRAINING_ROOT)
    except FileNotFoundError:
        return None, f"Training root not found: {TRAINING_ROOT}"
    except OSError as e:
        return None, f"Could not read training root '{TRAINING_ROOT}': {e}"

    dirs = []
    for entry in entries:
        full = os.path.join(TRAINING_ROOT, entry)
        if os.path.isdir(full):
            dirs.append(entry)
    dirs.sort()
    return dirs, None


def resolve_face_dir(user_id: str, dirs):
    if user_id in dirs:
        return user_id, "exact"
    case_matches = [entry for entry in dirs if entry.lower() == user_id.lower()]
    if len(case_matches) == 1:
        return case_matches[0], "case-insensitive"
    if len(case_matches) > 1:
        return None, "ambiguous"
    return None, "missing"


def run_retraining() -> int:
    try:
        result = subprocess.run(
            [sys.executable, TRAINER_SCRIPT],
            cwd=BASE_DIR,
            capture_output=True,
            text=True,
            check=False,
        )
        if result.stdout:
            print(result.stdout.strip())
        if result.stderr:
            print(result.stderr.strip())
        if result.returncode != 0:
            print(f"Retraining failed with exit code {result.returncode}")
            return 3
        print("Retraining completed.")
        return 0
    except Exception:
        print("Failed to start trainer:")
        traceback.print_exc()
        return 3


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: deleteFace.py <user_id>")
        return 2

    user_id = normalize_user_id(sys.argv[1])
    if not user_id:
        print("Empty user_id after normalization.")
        return 2

    dirs, list_error = list_training_dirs()
    if list_error:
        print(list_error)
        return 2

    matched_dir, match_type = resolve_face_dir(user_id, dirs)
    if matched_dir is None:
        print(f"No face directory match found for '{user_id}'.")
        if match_type == "ambiguous":
            print("Multiple case-insensitive matches exist; deletion is unsafe.")
        if dirs:
            sample = ", ".join(dirs[:MAX_SAMPLE_DIRS])
            more = len(dirs) - min(len(dirs), MAX_SAMPLE_DIRS)
            print(f"Available face directories ({len(dirs)} total): {sample}")
            if more > 0:
                print(f"...and {more} more")
        else:
            print("No face directories currently exist in training_images.")

        retrain_code = run_retraining()
        if retrain_code == 0:
            print("No matching folder was deleted, but model retraining completed to clear stale labels.")
            return 4
        return 1

    face_dir = os.path.join(TRAINING_ROOT, matched_dir)
    if match_type != "exact":
        print(f"Resolved '{user_id}' to folder '{matched_dir}' via {match_type} match.")

    try:
        shutil.rmtree(face_dir)
        print(f"Deleted user directory: {face_dir}")
    except PermissionError as e:
        print(f"Permission denied removing {face_dir}: {e}")
        return 2
    except FileNotFoundError:
        print(f"Directory vanished before delete: {face_dir}")
        retrain_code = run_retraining()
        return 4 if retrain_code == 0 else 1
    except OSError as e:
        print(f"OS error removing {face_dir}: {e}")
        return 2

    retrain_code = run_retraining()
    if retrain_code != 0:
        return 3
    return 0


if __name__ == "__main__":
    sys.exit(main())
